packages/ubuntu_desktop_installer/lib/pages/connect_to_internet/wifi_model.dart

```diff
@@ -0,0 +1,314 @@
+    log.debug('Enable wireless networking');
+    return _service.setWirelessEnabled(true);
+  }
+
+  @override
+  Future<void> connect() {
+    final device = selectedDevice!;
+    final accessPoint = device.selectedAccessPoint!;
+    log.debug('Connect $device to $accessPoint');
+    return _service.addAndActivateConnection(
+      connection: <String, Map<String, DBusValue>>{},
+      device: device.device,
+      accessPoint: accessPoint.accessPoint,
+    );
+  }
+
+  final NetworkService _service;
+  final UdevService? _udev;
+
+  List<WifiDevice>? _devices;
+  List<WifiDevice> get devices => _devices ??= _getDevices();
+
+  List<WifiDevice> _getDevices() {
+    final devices = _service.wirelessDevices
+        .map((device) => WifiDevice(device, _udev))
+        .toList();
+    log.debug(() => 'Update Wi-Fi devices: ${describeDevices(devices)}');
+    return devices;
+  }
+
+  void _resetDevices() {
+    if (_devices == null) return;
+    for (final device in _devices!) {
+      device.dispose();
+    }
+    _devices = null;
+  }
+
+  void _updateDevices() {
+    _resetDevices();
+    notifyListeners();
+  }
+
+  WifiDevice? _selected;
+  WifiDevice? get selectedDevice => _selected;
+  bool isSelectedDevice(WifiDevice device) => device == _selected;
+  void selectDevice(WifiDevice? device) {
+    selectedDevice?.removeListener(notifyListeners);
+    device?.addListener(notifyListeners);
+    _selected = device;
+    notifyListeners();
+  }
+
+  Future requestScan({String? ssid}) async {
+    if (!isEnabled) return;
+    final scans = <Future<void>>[];
+    for (final device in devices) {
+      if (device.isAvailable) {
+        scans.add(device.requestScan(ssid: ssid));
+      }
+    }
+    return Future.wait(scans);
+  }
+}
+
+class WifiDevice extends NetworkDevice {
+  WifiDevice(NetworkManagerDevice device, [UdevService? udev])
+      : _wireless = device.wireless!,
+        super(device, udev) {
+    addProperties(_wireless.propertiesChanged);
+    addPropertyListener('AccessPoints', _updateAccessPoints);
+    addPropertyListener('ActiveAccessPoint', notifyListeners);
+    addPropertyListener('LastScan', () {
+      _setLastScan(device.wireless!.lastScan);
+      _setScanning(false);
+      if (_completer?.isCompleted == false) {
+        _completer!.complete(device);
+      }
+    });
+  }
+
+  void init() => selectAccessPoint(activeAccessPoint);
+
+  @override
+  void dispose() {
+    _resetAccessPoints();
+    super.dispose();
+  }
+
+  @override
+  bool get isActive => super.isActive && activeAccessPoint != null;
+
+  bool get canConnect =>
+      isActive &&
+      _selected != null &&
+      _selected!.name == activeAccessPoint?.name;
+
+  final NetworkManagerDeviceWireless _wireless;
+
+  List<AccessPoint>? _accessPoints;
+  List<AccessPoint> get accessPoints => _accessPoints ??= _getAccessPoints();
+
+  List<AccessPoint> _getAccessPoints() {
+    final ssids = <List<int>>[];
+
+    bool hasSsid(List<int> ssid) {
```

Comment:
That's probably a change for the `nm` package, but I'll suggest it here while I'm at it: would it make sense to have a `SSID` type/class to hide away implementation details (like comparing two lists of integers, or converting from/to a string)?